otherlibs/systhreads/st_win32.h

```diff
@@ -24,7 +24,241 @@ Caml_inline void st_msleep(int msec)
+    InitializeConditionVariable(&m->is_free);
+    m->init = true;
+  }
+  m->busy = true;
+  atomic_store_release(&m->waiters, 0);
+  return 0;
+}
+
+static uintnat st_masterlock_waiters(st_masterlock * m)
+{
+  return atomic_load_acquire(&m->waiters);
+}
+
+static void st_masterlock_acquire(st_masterlock *m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  while (m->busy) {
+    atomic_fetch_add(&m->waiters, +1);
+    SleepConditionVariableSRW(&m->is_free, &m->lock,
+                              INFINITE, 0 /* exclusive */);
+    atomic_fetch_add(&m->waiters, -1);
+  }
+  m->busy = true;
+  st_bt_lock_acquire();
+  ReleaseSRWLockExclusive(&m->lock);
+
+  return;
+}
+
+static void st_masterlock_release(st_masterlock * m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  m->busy = false;
+  st_bt_lock_release(st_masterlock_waiters(m) == 0);
+  WakeConditionVariable(&m->is_free);
+  ReleaseSRWLockExclusive(&m->lock);
+
+  return;
+}
+
+/* Scheduling hints */
+
+/* This is mostly equivalent to release(); acquire(), but better. In particular,
+   release(); acquire(); leaves both us and the waiter we signal() racing to
+   acquire the lock. Calling yield or sleep helps there but does not solve the
+   problem. Sleeping ourselves is much more reliable--and since we're handing
+   off the lock to a waiter we know exists, it's safe, as they'll certainly
+   re-wake us later.
+*/
+Caml_inline void st_thread_yield(st_masterlock * m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  /* We must hold the lock to call this. */
+
+  /* We already checked this without the lock, but we might have raced--if
+     there's no waiter, there's nothing to do and no one to wake us if we did
+     wait, so just keep going. */
+  uintnat waiters = st_masterlock_waiters(m);
+
+  if (waiters == 0) {
+    ReleaseSRWLockExclusive(&m->lock);
+    return;
+  }
+
+  m->busy = false;
+  atomic_fetch_add(&m->waiters, +1);
+  WakeConditionVariable(&m->is_free);
+  /* Releasing the domain lock but not triggering bt messaging.
+     Messaging the bt should not be required because yield assumes
+     that a thread will resume execution (either the yielding thread
+     or a waiting thread). */
+  caml_release_domain_lock();
+
+  do {
+    /* Note: the POSIX spec prevents the above signal from pairing with this
+       wait, which is good: we'll reliably continue waiting until the next
+       yield() or enter_blocking_section() call (or we see a spurious condvar
+       wakeup, which are rare at best.) */
+       SleepConditionVariableSRW(&m->is_free, &m->lock,
+                                 INFINITE, 0 /* exclusive */);
+  } while (m->busy);
+
+  m->busy = true;
+  atomic_fetch_add(&m->waiters, -1);
+
+  caml_acquire_domain_lock();
+
+  ReleaseSRWLockExclusive(&m->lock);
+
+  return;
+}
+
+/* Triggered events */
+
+typedef struct st_event_struct {
+  SRWLOCK lock;                 /* to protect contents */
+  bool status;                  /* false = not triggered, true = triggered */
```

Comment:
@dustanddreams should this field also be marked atomic or volatile? There's a similar loop pattern in `st_event_wait`.

Reply:
Yes, I think so.